src/general/ActionHistory.cs

```diff
@@ -15,10 +15,10 @@ public class ActionHistory<T>
     ///   possibly redone), is 0 if there is none.
     /// </summary>
     [JsonProperty]
-    protected int ActionIndex { get; private set; }
+    public List<T> Actions { get; private set; } = new();
 
     [JsonProperty]
-    protected List<T> Actions { get; private set; } = new();
+    public int ActionIndex { get; private set; }
```

Comment:
These should not be made public.

Instead there should be a helper method to inspect the latest action. That way any code in the project cannot modify the action history however they like and end up with a massive mess.